Create new API chunking test

BobyMCbobs opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Adds a new API chunking test

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

hh commented:
/ok-to-test

devaips commented:
/retest

devaips commented:
/retest

devaips commented:
/area conformance

devaips commented:
/assign @timothysc

timothysc commented:
cc @kubernetes/sig-api-machinery-pr-reviews

timothysc commented:
@BobyMCbobs - New tests should go through the sig for addition. 
/unassign @timothysc 

I can make blanket changes to the framework, but this is sig-owned property.

lavalamp commented on test/e2e/apimachinery/chunking.go:
```diff
@@ -122,6 +122,29 @@ var _ = SIGDescribe("Servers with support for API chunking", func() {
+	ginkgo.It("should chunk lists of Pods", func() {
+		ns := f.Namespace.Name
+		c := f.ClientSet
+		client := c.CoreV1().PodTemplates(ns)
```
did you mean pods? otherwise the test title is wrong

liggitt replied:
we shouldn't use PodTemplates objects in new tests. see https://github.com/kubernetes/kubernetes/issues/82435

lavalamp commented on test/e2e/apimachinery/chunking.go:
```diff
@@ -122,6 +122,29 @@ var _ = SIGDescribe("Servers with support for API chunking", func() {
+		loopCount := int64(0)
+		for {
+			opts := metav1.ListOptions{}
+			opts.Limit = int64(rand.Int31n(numberOfTotalResources/10) + 1)
```
what if there is only 1? 0?

BobyMCbobs replied:
`loopCount` is the number of iterations that the `for {` loop has gone through

lavalamp commented on test/e2e/apimachinery/chunking.go:
```diff
@@ -122,6 +122,29 @@ var _ = SIGDescribe("Servers with support for API chunking", func() {
+			framework.ExpectNoError(err, "failed fetching PodTemplate list")
+
+			if loopCount > 0 && len(list.Continue) > 0 {
+				gomega.Expect(list.Continue[loopCount]).NotTo(gomega.Equal(list.Continue[loopCount-1]), "chunks are the same")
```
list.Continue is a string, right?

why are you using loop count as an index into this string?

the intention is to test that continue is different over different iterations of the loop? but that's not a valid assumption if e.g. you're on the last page.

caesarxuchao replied:
I have the same questions.

BobyMCbobs replied:
I have updated the test in cee471ad3a972b37fd928aa6af1891d8323b75df to store `list.Continue` in an array for comparison, since it itself isn't an array of `Continue`s.

lavalamp commented on test/e2e/apimachinery/chunking.go:
```diff
@@ -122,6 +122,29 @@ var _ = SIGDescribe("Servers with support for API chunking", func() {
+		client := c.CoreV1().PodTemplates(ns)
+		loopCount := int64(0)
+		for {
+			opts := metav1.ListOptions{}
```
Don't you need to set opts to use the continue token from the last list iteration?

BobyMCbobs replied:
I have updated the test to use what you suggested in https://github.com/ii/kubernetes/commit/cee471ad3a972b37fd928aa6af1891d8323b75df#diff-800e2a08a32f7d269d6af9e0c51c7d6dR137; It will only use it, if it's had it's first iteration.

lavalamp commented:
What is this test trying to test?

/assign @caesarxuchao

devaips commented:
@lavalamp this test was made in response to the outcome of PR https://github.com/kubernetes/kubernetes/pull/82297

The intention is to test that large request chunking is functioning. Reference: https://github.com/kubernetes/community/pull/896

caesarxuchao commented on test/e2e/apimachinery/chunking.go:
```diff
@@ -122,6 +122,29 @@ var _ = SIGDescribe("Servers with support for API chunking", func() {
+			loopCount++
+			e2elog.Logf("Fetched %d chunks; Continue: %s; Items: %s", loopCount, list.Continue, fmt.Sprintf("%T", list.Items))
+
+			if len(list.Continue) == 0 || loopCount == int64(len(list.Continue)) {
```
What does "loopCount == int64(len(list.Continue))" check?

BobyMCbobs replied:
This is a check to make sure it's at the end of the chunking list

caesarxuchao replied:
I think "len(list.Continue) == 0" means it's at the end of chunking list.

I still don't get what "loopCount == int64(len(list.Continue))" indicates. Could you elaborate?

BobyMCbobs replied:
The test has been updated in https://github.com/ii/kubernetes/commit/cee471ad3a972b37fd928aa6af1891d8323b75df and now no longer uses this section of code.

caesarxuchao commented:
> @lavalamp this test was made in response to the outcome of PR #82297

Seems a typo. Why did you link back to this PR itself?

> The intention is to test that large request chunking is functioning

I think it's covered by https://github.com/kubernetes/kubernetes/blob/38753be19d0f3da9e32c69751360169551d29ecb/test/e2e/apimachinery/chunking.go#L78-L123.

alejandrox1 commented:
/cc

devaips commented:
> > @lavalamp this test was made in response to the outcome of PR #82297
> 
> Seems a typo. Why did you link back to this PR itself?
> 

You're right! Here's the PR I intended to link to: https://github.com/kubernetes/kubernetes/pull/79871


> > The intention is to test that large request chunking is functioning
> 
> I think it's covered by
> 
> https://github.com/kubernetes/kubernetes/blob/38753be19d0f3da9e32c69751360169551d29ecb/test/e2e/apimachinery/chunking.go#L78-L123
> .

That was the test in question from that PR. Take a look at the comments. It was determined that the test wasn't suitable for conformance as is, due it its broadness. We are focusing on PodTemplate with this test.

However since making this test it looks like the PodTemplate API will be deprecated. See: https://github.com/kubernetes/kubernetes/issues/82435

So we may need to reevaluate this test altogether.

caesarxuchao commented:
I'm still confused. Are you referring to https://github.com/kubernetes/kubernetes/pull/79871#issuecomment-514854868? 

"if we want to exercise the behavior that is "api should chunk list calls" we need this specific test to be rewritten (or split into multiple test cases) for a variety of resources, not just PodTemplate"

Then why is this PR still testing PodTemplate?

BobyMCbobs commented:
/test pull-kubernetes-integration

lavalamp commented:
This is still using podtemplates, it shouldn't.

This will be super flaky because what's to prevent some other actor from modifying the collection?

The count of objects probably isn't enough, you need to verify that the exact same objects are present no matter the chunking. (once you solve the above problem)

This should probably test corner cases, like what happens if you insert an object right at / before / after the current continue token?

I don't think testing that the continue token is different every time is very useful. It probably will be but what user-facing thing is broken if it isn't?

devaips commented:
@lavalamp any suggestions on which object to use instead?

BobyMCbobs commented:
> This is still using podtemplates, it shouldn't.

> > @lavalamp any suggestions on which object to use instead?

I'm going to redo the references to PodTemplates in the whole file has been updated to use ConfigMaps

> 
> This will be super flaky because what's to prevent some other actor from modifying the collection?
> 

@lavalamp You may be correct about the flakiness of the test results, but it may be best to just wait two weeks to confirm that.

> The count of objects probably isn't enough, you need to verify that the exact same objects are present no matter the chunking. (once you solve the above problem)

In https://github.com/kubernetes/kubernetes/pull/82297/commits/f26e0dfa2e09ae6fb5841e1e69c699d0544d2018, I've added labels to the `PodTemplate.ObjectMeta`, so the Pods with the generated label will only be retrieved.

smarterclayton commented on test/e2e/apimachinery/chunking.go:
```diff
@@ -122,6 +129,61 @@ var _ = SIGDescribe("Servers with support for API chunking", func() {
 		gomega.Expect(list.Items).To(gomega.HaveLen(numberOfTotalResources))
 	})
 
+	ginkgo.It("should chunk lists of Pods", func() {
```
Why are we adding a new test instead of just updating the existing ones to use config maps?  What does this new test add?

smarterclayton commented:
I disagree with the referenced PR - we should switch the existing tests to not use pod templates, and then extend function.

We should then add testing for a CRD type to verify the behavior is consistent.  We don’t need to verify chunking for more than one built in type.

BobyMCbobs commented:
This PR was updated in https://github.com/kubernetes/kubernetes/pull/82297/commits/4184e8f3f40734f1ff75d55866344d6061989878 to change the resource type from PodTemplates to ConfigMaps. 

The first test in the file had to be updated and works as it did prior.
The slow test at the end of the file was broken and considered unneeded or a more specific test should be written for checking that the continue string tokens expires as it should.

BobyMCbobs commented:
/retest

Commit:
Make chunking tests use ConfigMaps instead of PodTemplates

This also replace an old [slow] test which did not work with ConfigMaps
as it were.

Co-Authored-By: Caleb Woodbine <caleb@ii.coop>

Commit:
chunking.go: Replace e2elog.Logf with framework.Logf

Co-Authored-By: Caleb Woodbine <caleb@ii.coop>

BobyMCbobs committed:
Update: gomega.Expect to framework.Expect -- also add failure messages

BobyMCbobs commented:
/test pull-kubernetes-kubemark-e2e-gce-big

devaips commented:
/assign @timothysc

timothysc commented:
> I disagree with the referenced PR - we should switch the existing tests to not use pod templates, and then extend function.  We should then add testing for a CRD type to verify the behavior is consistent. We don’t need to verify chunking for more than one built in type.

@devaips - I agree with Clayton's advice ^ .

BobyMCbobs commented:
> I disagree with the referenced PR - we should switch the existing tests to not use pod templates, and then extend function.
> 
> We should then add testing for a CRD type to verify the behavior is consistent. We don’t need to verify chunking for more than one built in type.

---

> > I disagree with the referenced PR - we should switch the existing tests to not use pod templates, and then extend function.  We should then add testing for a CRD type to verify the behavior is consistent. We don’t need to verify chunking for more than one built in type.
> 
> @devaips - I agree with Clayton's advice ^ .

Just to clarify, @smarterclayton are you meaning for the test to be modular, where the object kind can be swapped out or iterated through the given types (which includes a CRD)? or are you meaning so that a separate test which tests API chunking for a CRD should be written?

smourapina commented:
@BobyMCbobs Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this PR still intended for 1.17?

BobyMCbobs commented:
> @BobyMCbobs Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this PR still intended for 1.17?

Yes I'd like it to be in for 1.17, but if it doesn't work out I'd like it to be in for 1.18.

hh commented:
/area conformance

josiahbjorgaard commented:
/milestone v1.18

BobyMCbobs commented:
/assign @caesarxuchao

spiffxp commented:
@BobyMCbobs re: https://github.com/kubernetes/kubernetes/pull/82297#issuecomment-541915349
I interpret as s/PodTemplate/ConfigMap in one PR, then template into something that can be used for ConfigMap and a CRD in another PR, and then extend to catch the corner cases @lavalamp describes in https://github.com/kubernetes/kubernetes/pull/82297#issuecomment-534243003

This PR seems to be changing one of the tests as well as s/PodTemplate/ConfigMap and I'm not clear why. Can we isolate the change in this PR to just moving to ConfigMap?

BobyMCbobs commented:
/check-cla

hh commented:
Closing via sig-arch conformance subproject meeting:
@BobyMCbobs to write up two mock-tests Issues based on the two different concepts we tried to merge here.

hh commented:
/close
